core/src/main/scala/org/bitcoins/core/wallet/utxo/AddressTag.scala

```diff
@@ -0,0 +1,74 @@
+package org.bitcoins.core.wallet.utxo
+
+sealed trait AddressTagType {
+  def tagType: String
+}
+
+sealed trait AddressTag {
+  def name: String
+  def tagType: String
+
+  def ==(at: AddressTag): Boolean = name == at.name && tagType == at.tagType
+  def !=(at: AddressTag): Boolean = !(this == at)
+
+  def toCustom: CustomAddressTag = CustomAddressTag(name, tagType)
+}
+
+trait AddressTagFactory[TagType <: AddressTagType] {
+
+  def tagType: String
+
+  def all: Vector[TagType]
+
+  def fromString(str: String): Option[TagType] = {
+    all.find(tag => str.toLowerCase() == tag.toString.toLowerCase)
+  }
+
+  def fromCustom(custom: CustomAddressTag): Option[TagType] = {
+    // Allow empty tagType because sometimes (like in the db mappers) we don't have access to
+    // the tagType when we are getting the tag, so we need to be able to pass it in.
+    require(
+      custom.tagType.isEmpty || custom.tagType == tagType,
+      s"If tag's tagType is defined it must be the same as tagType, got ${custom.tagType}")
+
+    fromString(custom.name)
+  }
+}
+
+sealed trait InternalAddressTagType extends AddressTagType
+sealed trait InternalAddressTag extends AddressTag
+
+trait ExternalAddressTagType extends AddressTagType
+trait ExternalAddressTag extends AddressTag
+
+case class CustomAddressTag(name: String, tagType: String)
+    extends ExternalAddressTag
```

Comment:
I'm not sure I understand the point of having `CustomAddressTag` extend `ExternalAddressTag`? Isn't the point of having internal and external that this is no longer needed? Or perhaps you still want an internal representation for external tags in which case this should extend `InternalAddressTag` and should probably be renamed to `UnknownAddressTag` or something like this

Reply:
>  perhaps you still want an internal representation for external tags

Yes, this is the main need of this. I like `UnknownAddressTag` better so I will do that.